### PR TITLE
[Feature] PILCO

### LINKED ISSUE #3513: [Feature Request] Implement PILCO (Probabilistic Inference for Learning Control)
## Motivation

Implement PILCO (Probabilistic Inference for Learning Control), as requested in #509.

## Solution

It uses Gaussian Processes to model dynamics and analytic moment matching to propagate uncertainty, allowing for direct gradient-based policy optimization.

## Alternatives

NA

## Additional context

Reference: [PILCO: A Model-Based and Data-Efficient Approach to Policy Search](https://www.google.com/search?q=https://ieeexplore.ieee.org/document/6130309)

## Checklist

- [X] I have checked that there is no similar issue in the repo (**required**)

### PR DESCRIPTION
**Pull request by PSXBRosa**
## Description

This PR introduces the implementation of the PILCO (Probabilistic Inference for Learning Control) algorithm to TorchRL.

Key details of the implementation:

- Gaussian Process Regression: I utilized the external libraries botorch and gpytorch for the GPR components. this avoids the overhead and complexity of maintaining a custom GPR implementation.
- Moment Matching: I initially considered and experimented with a Monte Carlo approach for moment matching to simplify the underlying mathematics. While I couldn't get the MC approach to stabilize and work correctly (though it remains an interesting alternative), the current working implementation relies on the analytical forms for moment matching. This aligns directly with what was done in the original PILCO paper by Deisenroth and Rasmussen.
- Credits: I want to give a huge thanks and credit to @nrontsis. I used the code from their repository (https://github.com/nrontsis/PILCO) as a valuable implementation reference and to test/validate different parts of my own PyTorch adaptation.

## Motivation and Context

PILCO is a highly sample-efficient model-based reinforcement learning algorithm, making it a valuable addition to the library's algorithm suite.

close #3513

- [x] I have raised an issue to propose this change ([required](https://github.com/pytorch/rl/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [x] New feature (non-breaking change which adds core functionality)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: sota-implementations/pilco/utils.py [hunk 1]
```diff
@@ -0,0 +1,585 @@
...
+
+    v = torch.matmul(U, diff)
+    tmp = torch.cholesky_solve(v, L)
+    quad = torch.matmul(v.transpose(-2, -1), tmp)
+    exp_term = (-0.5 * quad).squeeze(-1).squeeze(-1)
+
+    return (1.0 - det_term * torch.exp(exp_term)).sum(dim=1)
+
+
+class BoTorchGPWorldModel(nn.Module):
```
Comment (vmoens) [inline sota-implementations/pilco/utils.py]: If properly documented i'm happy with having this in core!

File: sota-implementations/pilco/utils.py [hunk 2]
```diff
@@ -0,0 +1,585 @@
...
+        delta_std_flat = torch.stack(stds, dim=-1)
+
+        batch_shape = observation_mean.shape[:-1]
+        delta_mean = delta_mean_flat.view(*batch_shape, self.obs_dim)
+        delta_std = delta_std_flat.view(*batch_shape, self.obs_dim)
+
+        return observation_mean + delta_mean, torch.diag_embed(delta_std**2)
+
+
+class ImaginedEnv(ModelBasedEnvBase):
```
Comment (vmoens) [inline sota-implementations/pilco/utils.py]: Ditto, maybe we want this in core.
How different is it from DreamerEnv? Can we blend the two together? (ok if we want to keep them separated)

File: sota-implementations/pilco/utils.py [hunk 3]
```diff
@@ -0,0 +1,585 @@
...
+            device=self.device,
+        )
+
+        out.set("done", torch.zeros(B, 1, dtype=torch.bool, device=self.device))
+        out.set("terminated", torch.zeros(B, 1, dtype=torch.bool, device=self.device))
+
+        return out
+
+
+class RBFController(nn.Module):
```
Comment (vmoens) [inline sota-implementations/pilco/utils.py]: ditto happy to have it in core

File: sota-implementations/pilco/utils.py [hunk 4]
```diff
@@ -0,0 +1,585 @@
...
+            batch_size, self.obs_dim, self.obs_dim, dtype=m_x.dtype, device=m_x.device
+        )
+
+        X_i = X_train.unsqueeze(1)
+        X_j = X_train.unsqueeze(0)
+        diff = X_i - X_j
+        joint_mean_flat = joint_mean.unsqueeze(1).unsqueeze(1)
+
+        for a in range(self.obs_dim):
+            for b in range(self.obs_dim):
```
Comment (vmoens) [inline sota-implementations/pilco/utils.py]: a lot in here can be vectorized

File: sota-implementations/pilco/utils.py [hunk 5]
```diff
@@ -0,0 +1,585 @@
...
+        if x_var is not None:
+            observation_uncertain = not torch.all(
+                torch.isclose(x_var, torch.zeros_like(x_var))
+            )
+        if observation_uncertain:
+            return self.uncertain_forward(action, observation)
+        else:
+            return self.deterministic_forward(action, observation)
+
+    def freeze_and_detach(self) -> None:
```
Comment (vmoens) [inline sota-implementations/pilco/utils.py]: is this needed?

File: sota-implementations/pilco/utils.py [hunk 6]
```diff
@@ -0,0 +1,585 @@
...
+
+    def compute_factorizations(self) -> tuple[torch.Tensor, torch.Tensor]:
+        return self._cached_inv_K, self._cached_beta
+
+    def _gather_gp_params(self) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        return self.lengthscales, self.variances, self.noises
+
+    def forward(
+        self, action: TensorDictBase, observation: TensorDictBase
+    ) -> tuple[torch.Tensor, torch.Tensor]:
```
Comment (vmoens) [inline sota-implementations/pilco/utils.py]: usually inputs in torchrl are a single tensordict

It should work as a standard TensorDictModule with in_keys/out_keys.

File: sota-implementations/pilco/utils.py [hunk 7]
```diff
@@ -0,0 +1,585 @@
...
+        action_mean = action.get("mean")
+
+        x_flat = observation_mean.view(-1, self.obs_dim)
+        u_flat = action_mean.view(-1, self.action_dim)
+
+        X_test = torch.cat([x_flat, u_flat], dim=-1)
+
+        means, stds = [], []
+
+        with torch.no_grad():
```
Comment (vmoens) [inline sota-implementations/pilco/utils.py]: needed? Maybe make parametrizable?

File: sota-implementations/pilco/utils.py [hunk 8]
```diff
@@ -0,0 +1,585 @@
...
+
+        self.centers = nn.Parameter(torch.randn(n_basis, input_dim) * 0.5)
+        self.weights = nn.Parameter(torch.randn(n_basis, output_dim) * 0.1)
+        self.lengthscales = nn.Parameter(torch.ones(input_dim))
+        self.variance = 1.0
+
+    @staticmethod
+    def squash_sin(
+        m: torch.Tensor, s: torch.Tensor, max_action: float | torch.Tensor
+    ) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
```
Comment (vmoens) [inline sota-implementations/pilco/utils.py]: non-trivial, good work!
Should be brought to the core lib and tested!

File: sota-implementations/pilco/utils.py [hunk 9]
```diff
@@ -0,0 +1,585 @@
...
+    """Creates the transformed environment."""
+    env = TransformedEnv(
+        GymEnv(env_name, pixels_only=False, from_pixels=from_pixels, device=device)
+    )
+    env.append_transform(RewardSum())
+    env.append_transform(StepCounter())
+    return env
+
+
+def pendulum_cost(
```
Comment (vmoens) [inline sota-implementations/pilco/utils.py]: This is interesting but conflates two things: (a) a generic "saturating exponential cost under Gaussian uncertainty" formula, and (b) pendulum-specific defaults (4D identity weight matrix, zero target). In core, this should be:

- A proper `nn.Module` (e.g., `SaturatingCost` or `ExponentialQuadraticCost`) with configurable weights and target as constructor arguments.
- The pendulum-specific defaults would stay in the sota-implementation.
- This pattern of "cost functions that operate on belief states (mean, variance)" is general to all moment-matching MBRL methods.

File: sota-implementations/pilco/utils.py [hunk 10]
```diff
@@ -0,0 +1,585 @@
...
+                    torch.matmul(beta[a].unsqueeze(0), Qb.unsqueeze(-1))
+                    .squeeze(-1)
+                    .squeeze(-1)
+                )
+
+                if a == b:
+                    invK_Q = torch.matmul(inv_K[a].unsqueeze(0), Q_ab)
+                    trace_val = torch.diagonal(invK_Q, dim1=-2, dim2=-1).sum(-1)
+
+                    pred_cov[:, a, a] += variances[a] - trace_val + noises[a].item()
```
Comment (vmoens) [inline sota-implementations/pilco/utils.py]: avoid .item()
Call `.tolist()` if absolutely necessary before. I think a plain tensor should work. `.item()` breaks compile and requires cuda sync.

File: sota-implementations/pilco/utils.py [hunk 11]
```diff
@@ -0,0 +1,585 @@
...
+        B_mat = iL_batch @ S @ iL_batch + torch.eye(
+            D, device=device, dtype=m.dtype
+        ).unsqueeze(0)
+
+        iN = inp @ iL
+
+        t = torch.linalg.solve(B_mat, iN.mT).mT
+
+        exp_term = torch.exp(-0.5 * torch.sum(iN * t, dim=-1))
+        detB = torch.linalg.det(B_mat)
```
Comment (vmoens) [inline sota-implementations/pilco/utils.py]: torch.linalg.slogdet would be numerically more stable (and is already partially used via the Cholesky log-det pattern elsewhere in the same code)

File: sota-implementations/pilco/utils.py [hunk 12]
```diff
@@ -0,0 +1,585 @@
...
+        B_mat = B_mat + torch.eye(
+            self.input_dim, dtype=m_x.dtype, device=m_x.device
+        ).view(1, 1, self.input_dim, self.input_dim)
+
+        t = torch.linalg.solve(B_mat, inv_N.transpose(-2, -1)).transpose(-2, -1)
+
+        scaled_exp = torch.exp(-torch.sum(inv_N * t, dim=-1) / 2)
+        lb = scaled_exp * beta.unsqueeze(0)
+
+        det_B = torch.linalg.det(B_mat)
```
Comment (vmoens) [inline sota-implementations/pilco/utils.py]: ditto - let's think in logs!

File: sota-implementations/pilco/utils.py [hunk 13]
```diff
@@ -0,0 +1,585 @@
...
+                M_ab = joint_var + torch.diag(Lambda_ab).unsqueeze(0)
+
+                solved_z_flat = torch.linalg.solve(
+                    M_ab, z_flat.transpose(-2, -1)
+                ).transpose(-2, -1)
+                exp2 = (-0.5 * torch.sum(z_flat * solved_z_flat, dim=-1)).view(
+                    batch_size, num_train_pts, num_train_pts
+                )
+
+                det_R_ab = torch.linalg.det(R_ab)
```
Comment (vmoens) [inline sota-implementations/pilco/utils.py]: ditto let's think in logs!

File: sota-implementations/pilco/utils.py [hunk 14]
```diff
@@ -0,0 +1,585 @@
+from collections.abc import Sequence
+
+import torch
+import torch.nn as nn
+from botorch.fit import fit_gpytorch_mll
+
+from botorch.models import ModelListGP, SingleTaskGP
```
Comment (vmoens) [inline sota-implementations/pilco/utils.py]: botorch / gpytorch need to be added to the repo's dependencies as optional deps

**Comment by alektebel** [comment]
Hi @PSXBRosa ,

Thank you for your work implementing PILCO for torchrl. 

A few days ago I opened a discussion about adding MC-PILCO, discussion n 3538, and reading through @vmoens ' review it's clear we're both going to depend on the same core primitives (`BoTorchGPWorldModel`, `RBFController`). Since @vmoens is already asking for those to be moved into torchrl/modules/models/, it seems like the cleanest path forward is to coordinate so we don't end up duplicating changes or creating interface conflicts.

A couple of options as I see it:
1. We collaborate on this PR directly, so the core primitives are designed with both PILCO and MC-PILCO in mind from the start (e.g. making sure `BoTorchGPWorldModel` exposes both `uncertain_forward` and a `sample_forward` path for the MC variant).
2. I open a separate MC-PILCO PR after this one merges, building on top of your primitives once they're in main.

Happy to go with whatever works best for you. If you want to discuss further feel free to reply here or reach out on discord (cabesamotora) !

**Comment by PSXBRosa** [comment]
Hi @alektebel, thanks for reaching out.

#1 seems like the cleanest approach. I'm in the middle of a move right now so I haven't had much time for the PR this week, but I've started on vmoens' comments and plan to have the current issues resolved by the end of next week.

My only progress so far is moving the loss to core. I can push this as a WIP if you'd like to take a look? How do you envision building on top of the current classes? Do you have a specific inheritance plan in mind? I'll ping you on discord.

**Comment by vmoens** [comment]
@PSXBRosa I can help porting stuff to core if you want me to :)

**Comment by PSXBRosa** [comment]
@vmoens, I think that would help! I ended up having less time than I expected to work on the PR this last week.

File: torchrl/modules/models/gp.py [hunk 15]
```diff
@@ -0,0 +1,383 @@
...
+            batch_size, self.obs_dim, self.obs_dim, dtype=m_x.dtype, device=m_x.device
+        )
+
+        X_i = X_train.unsqueeze(1)
+        X_j = X_train.unsqueeze(0)
+        diff = X_i - X_j
+        joint_mean_flat = joint_mean.unsqueeze(1).unsqueeze(1)
+
+        for a in range(self.obs_dim):
+            for b in range(self.obs_dim):
```
Comment (vmoens) [inline torchrl/modules/models/gp.py]: let's optimize this when we can!

File: torchrl/modules/models/gp.py [hunk 16]
```diff
@@ -166,60 +198,53 @@ def _gather_gp_params(self) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
         """Returns the extracted hyperparameters of each per-dimension GP."""
         return self.lengthscales, self.variances, self.noises
 
-    def forward(
-        self, action: TensorDictBase, observation: TensorDictBase
-    ) -> tuple[torch.Tensor, torch.Tensor]:
+    def forward(self, tensordict: TensorDictBase) -> TensorDictBase:
```
Comment (PSXBRosa) [inline torchrl/modules/models/gp.py]: Based on a previous comment in the PR, I thought changing this method signature would be better.

File: sota-implementations/pilco/pilco.py [hunk 17]
```diff
@@ -119,6 +113,15 @@ def pilco_loop(
...
 
+        test_rollout.set("observation", test_rollout.get(("observation", "mean")))
+        test_rollout.set("action", test_rollout.get(("action", "mean")))
+        test_rollout.set(
+            ("next", "observation"), test_rollout.get(("next", "observation", "mean"))
+        )
+
+        test_rollout = test_rollout.select(
+            *rollout.keys(include_nested=True, leaves_only=True)
+        )
```
Comment (PSXBRosa) [inline sota-implementations/pilco/pilco.py]: Due to the changes to the GP class, this was necessary. I'm not sure if there's a cleaner way to handle this.

File: torchrl/modules/models/gp.py (hunk 15 above)
Comment (PSXBRosa) [reply torchrl/modules/models/gp.py]: sounds good! I'll work on this after work today

**Comment by PSXBRosa** [comment]
hey @vmoens, just wanted a quick review here to know what the next steps should be!

**Comment by vmoens** [comment]
image

@PSXBRosa 
Cleaning up and merging soon!

**Comment by alektebel** [comment]
Apologies for the delay, 

seeing that this work is about to get merged, I will add MC-PILCO in another PR in order to not stop this
